File: 39-combination_sum.py
# time: O(N^(T/M)) - number of nodes in the execution tree
# space: O(T/M) - worst case call stack: keep on adding the smallest element M until we reach target T
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:

        results = []

        def backtrack(remain, comb, start):
            if remain == 0:
                # make a deep copy of the current combination
                results.append(list(comb))
                return
            elif remain < 0:
                # exceed the scope, stop exploration.
                return

            for i in range(start, len(candidates)):
                # add the number into the combination
                comb.append(candidates[i])
                # give the current number another chance, rather than moving on
                backtrack(remain - candidates[i], comb, i)
                # backtrack, remove the number from the combination
                comb.pop()

        backtrack(target, [], 0)

        return results

File: 39-combination_sum.cs
public class Solution {
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        List<IList<int>> combos = new();
        backtrack(0, target, new());
        return combos;

        void backtrack(int start, int remain, Stack<int> combo) {
            if (remain == 0) {
                combos.Add(new List<int>(combo));
                return;
            }
            if (remain < 0) {
                return;
            }

            for (int i = start; i < candidates.Length; i++) {
                combo.Push(candidates[i]);
                backtrack(i, remain - candidates[i], combo);
                combo.Pop();
            }

        }

    }
}
